api/handlers.go

```diff
@@ -92,22 +110,30 @@ func (api *API) loginGET(w http.ResponseWriter, req *http.Request, _ httprouter.
 
 // loginPOST starts a user session by issuing a cookie
 func (api *API) loginPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	defer api.closeBody(req)
+	// Get the body, we might need to use it several times.
```

Comment:
you're not using it several times right? I think this is redundant as you're more or less doing:
```
- body := ioutil.ReadAll(req.Body)
- buffer := bytes.NewBuffer(body)
- body := ioutil.ReadAll(buffer)
```

Reply:
It's because it's not necessarily a minor performance hit. I can send a very large body and you would be making it easy for me because you would be allocating twice the amount of memory. You mention you need an `io.Reader` in another, are you referring to `LoadFromReader`? Because there the first thing we do is `b, err := ioutil.ReadAll(body)`, perhaps we need `LoadFromBytes` to go alongside it? Or a json decoder is probably best.

Reply:
The `LoadFromBytes` is a good idea, thanks.

A JSON decoder can be helpful if we have a very long stream of JSON objects but it won't protect us from the malicious body. What will protect us is a limit reader. I'll write one and we can discuss whether it works well for the purpose.

Reply:
* `LoadFromBytes` implemented
* `io.LimitedReader` user for reading request bodies

Reply:
f/u I would scan the codebase for `ioutil.ReadAll` and then just get rid of all of them. It's usually not something you want to use except for testing since you can always just pass the limited reader to a `json.Decoder`. That's also more efficient since it will decode the body as it comes in without reading the whole thing first and then decoding it.